freqtrade/optimize/backtesting.py

```diff
@@ -350,8 +351,70 @@ def _get_close_rate(self, sell_row: Tuple, trade: LocalTrade, sell: SellCheckTup
         else:
             return sell_row[OPEN_IDX]
 
+    def _get_adjust_trade_entry_for_candle(self, trade: LocalTrade, row: Tuple
+                                           ) -> LocalTrade:
+
+        current_profit = trade.calc_profit_ratio(row[OPEN_IDX])
+
+        stake_amount = strategy_safe_wrapper(self.strategy.adjust_trade_position,
+                                             default_retval=None)(
+            pair=trade.pair, trade=trade, current_time=row[DATE_IDX].to_pydatetime(),
+            current_rate=row[OPEN_IDX], current_profit=current_profit)
+
+        # Check if we should increase our position
+        if stake_amount is not None and stake_amount > 0.0:
+            return self._execute_trade_position_change(trade, row, stake_amount)
+
+        return trade
+
+    def _execute_trade_position_change(self, trade: LocalTrade, row: Tuple,
+                                       stake_amount: float) -> LocalTrade:
+        current_price = row[OPEN_IDX]
+        propose_rate = min(max(current_price, row[LOW_IDX]), row[HIGH_IDX])
+        available_amount = self.wallets.get_available_stake_amount()
+
+        try:
+            min_stake_amount = self.exchange.get_min_pair_stake_amount(
+                trade.pair, propose_rate, -0.05) or 0
+            stake_amount = self.wallets.validate_stake_amount(trade.pair,
+                                                              stake_amount, min_stake_amount)
+            stake_amount = self.wallets._check_available_stake_amount(stake_amount,
+                                                                      available_amount)
+        except DependencyException:
+            logger.debug(f"{trade.pair} adjustment failed, "
+                         f"wallet is smaller than asked stake {stake_amount}")
+            return trade
+
+        amount = stake_amount / current_price
+        if amount <= 0:
+            logger.debug(f"{trade.pair} adjustment failed, amount ended up being zero {amount}")
+            return trade
+
+        buy_order = Order(
+            ft_is_open=False,
+            ft_pair=trade.pair,
+            symbol=trade.pair,
+            ft_order_side="buy",
+            side="buy",
+            order_type="market",
+            status="closed",
+            price=propose_rate,
+            average=propose_rate,
+            amount=amount,
+            cost=stake_amount
+        )
+        trade.orders.append(buy_order)
+        trade.recalc_trade_from_orders()
+        self.wallets.update()
```

Comment:
I'm currently hesitant / undecided if this is needed or not.

i don't THINK it is (the main method is `get_trade_stake_amount` - which has this builtin). Please look at the other comments for this file first - it may "resolve" itself automatically...

Reply:
Reworked it all into _enter_trade, so no more code duplication. recalc is called only for pos_adjust and wallets update can be skipped here indeed